common/src/main/java/org/conscrypt/ConscryptEngineSocket.java

```diff
@@ -447,14 +449,22 @@ public final void close() throws IOException {
         }
 
         try {
-            // Close the underlying socket.
-            super.close();
-        } finally {
             // Close the engine.
             engine.closeInbound();
             engine.closeOutbound();
```

Comment:
we are potentially calling closeOutbound() twice, can we draining before this line?

Reply:
That's fine it's idempotent.  First call closes the SSL session and queue a close alert.  drainOutputQueue unqueues that from the SSLEngine and queues it for transmission on the socket.  The second closeOutbound call notes that the socket is now fully closed with no remaining data to be sent and releases resources.

However I'm still not happy about this, as we could potentially block for a long time in drainOutputQueue.  We could potentially do that on a background thread or executor, but if the app is experiencing network problems we might be queueing a lot of threads or tasks which is obviously bad.

Note that the structure in this PR isn't worse than the fd-based socket.  That calls BoringSSL's SSL_close() which also tries to send the close notify and so can potentially block.